rfcs/20250312-ukernel-api-updates/README.md

```diff
@@ -0,0 +1,262 @@
+
+    // Check for transform requirements.
+    if (pack_type != dnnl::ukernel::pack_type::no_trans) {
+        // Additional handling of transform.
+        ...
+        dnnl::ukernel::transform tr(..., /* allow_empty = */ true);
+        if (!tr) {
+            printf("API can't be used\n");
+            fallback_call();
+        }
+    }
+
+    ...
+
+    dnnl::ukernel::brgemm brg(..., /* allow_empty = */ true);
+    // Rare case but safer to check.
+    if (!brg) {
+        printf("API can't be used\n");
+        fallback_call();
+    }
+
+    // Configurating `brg` object...
+
+    // This is possible with proposed API change.
+    const bool ok = brg.finalize();
+    if (!ok) {
+        printf("API can't be used\n");
+        fallback_call();
+    }
+
+    ...
+```
+
+### Runtime Leading Dimension Support
+
+Some applications programming model might not have all the shapes information
+available at the kernel creation time. The strategy for such applications is
+to generate a pre-defined set of "tiles", let's for the sake of simplicity
+announce two: 16x16 and 32x32 (though there can be more), which can work with
+arbitrary leading dimension values provided at runtime. The example of
+application is [triton-cpu](https://github.com/triton-lang/triton-cpu), which is
+currently a fork of the original triton.
+
+The library already provides a notion of runtime dimension values through the
+special `DNNL_RUNTIME_DIM_VAL` macro value. The only part that is not there is
+passing leading dimensions at runtime.
+
+The proposal is to update the `execute` with extra argument with a default empty
+value:
+
+```cpp
+/// dnnl_ukernel.hpp
+
+/// BRGeMM ukernel
+struct brgemm : public handle<dnnl_brgemm_t> {
+    ...
+
+    /// Executes a BRGeMM ukernel object.
+    ///
+    /// @param A Base pointer to a tensor A.
+    /// @param B Base pointer to a tensor B.
+    /// @param A_B_offsets Vector of pairs of tensors A and B offsets for
+    ///     each batch. The number of batches must coincide with the
+    ///     `batch_size` value passed at object construction stage.
+    /// @param C Pointer to a tensor C (accumulation buffer).
+    /// @param scratchpad Pointer to a scratchpad buffer.
+    /// @param actual_lds NEW ARGUMENT. Vector of actual leading dimensions for
+    ///     tensors A, B, and C. Must be specified if brgemm object was created
+    ///     with `DNNL_RUNTIME_DIM_VAL` instead of regular values for `lda`,
+    ///     `ldb`, or `ldc`. When `get_B_pack_type` returns values other than
+    ///     `no_trans`, `actual_lds[1]` for `ldb` must be used in transform
+    ///     routine. If any of given values were not specified as runtime values
+    ///     at kernel creation, they will be used instead.
+    void execute(const void *A, const void *B,
+            const std::vector<std::pair<memory::dim, memory::dim>> &A_B_offsets,
+            void *C, void *scratchpad,
+            const std::vector<memory::dim> &actual_lds = {}) const {
```

Comment:
Given that:
- there could be potential overheads due to heap allocation in std::vector
- and that we need only `lda`, `ldb`, `ldc`, `ldd`.

Would it make sense to actually take `lda`, `ldb`, `ldc`, `ldd` separately ? Or use `const std::array<memory::dim, 4> &lds` ?

Reply:
It would also avoid potential out of bounds accesses and make their internal use easier.